website/blog/2024-03-18-deep-learning-part-2/index.md

```diff
@@ -0,0 +1,201 @@
+
+![](finetune_with_MLflow.png)
+
+Using MLflow while fine tuning ML model in Databricks
+
+
+# The Importance of Logging
+
+The integration of the Hugging Face trainer callback with MLFlow plays a pivotal role in this ecosystem. It allows for comprehensive tracking and logging of metrics, parameters, and artifacts during model finetuning. This logging can be configured flexibly – at every epoch, after a set number of epochs, or steps. Importantly, it can cover the entire model checkpoint or be limited to just the adapter when using [PEFT (Parameter-Efficient Fine-Tuning)](https://huggingface.co/docs/peft/index) methods like [QLoRA](https://arxiv.org/abs/2305.14314). We have highlighted the additional code necessary for MLflow to log every epoch and checkpoint.
+
+
+## Configuring Hugging Face Trainer Callback for MLflow
+
+```python
+class CustomMLflowCallback(MLflowCallback):
+    """
+    A [`TrainerCallback`] that sends the logs to [MLflow](https://www.mlflow.org/). Can be disabled by setting
+    environment variable `DISABLE_MLFLOW_INTEGRATION = TRUE`.
+    """
+
+  def on_save(self, args, state, control,model,tokenizer, **kwargs):
+        if self._initialized and state.is_world_process_zero and self._log_artifacts:
+           ckpt_dir = f"checkpoint-{state.global_step}"
+  	 self._ml_flow.transformers.log_model(
+                             transformers_model={
+                                 "model": model,
+                                 "tokenizer": tokenizer,
+                             },
+                             task="text-generation",
+                             artifact_path=ckpt_dir,
+                             pip_requirements=["torch", "transformers", "accelerate"],
+                             input_example=input_example,
+                             signature=signature,
+                             metadata={"task": "llm/v1/completions"}
+                         )
+```
+
+See detailed code block [here](https://github.com/puneet-jain159/fine_tuning_llm_with_mlflow_blog/blob/495c156f73435996706c42bfda4bb7519dc21dab/databricks_llm/callbacks.py#L64)
+
+In the code snippet above, we create a new callback class that includes an additional method, **on_save**, which logs the weights and tokenizer after **every n** steps/epochs. This frequency is configured as part of the training arguments using the Transformer library. Additionally, we can employ a custom pyfunc to perform the same activity. For an example, please refer to the link provided [here](https://github.com/puneet-jain159/fine_tuning_llm_with_mlflow_blog/blob/main/databricks_llm/callbacks.py).
+
+The custom callback gracefully saves the model checkpoints during the finetuning process. The mlflow\.transformers.log\_model expression saves the model and the tokenizer alongside the dependencies, input example and the model signature during each logging event.
+
+
+## Scaling and Platform Considerations for Training
```

Comment:
This is a subsection of "The Importance to Logging"—is it supposed to be? I don't see how it relates.

Reply:
Not sure I undestand the concern